### DISCUSSION TITLE
[@types/node] annotation for NodeJS.Process.on(...）is not helpful, why?

### DISCUSSION THREAD

**Discussion by qiulang** [DefinitelyTyped/DefinitelyTyped#61108] [Issues with a @types package]
I see `process.d.ts` has following annotation

```
NodeJS.Process.on(event: NodeJS.Signals, listener: NodeJS.SignalsListener）
on(event: 'uncaughtExceptionMonitor', listener: UncaughtExceptionListener): this;
on(event: 'unhandledRejection', listener: UnhandledRejectionListener): this;
on(event: 'warning', listener: WarningListener): this;
on(event: 'message', listener: MessageListener): this;

type UnhandledRejectionListener = (reason: unknown, promise: Promise<unknown>) => void;
type WarningListener = (warning: Error) => void; 
type NodeJS.Signals = "SIGTERM" | "SIGABRT" | ... 
type SignalsListener = (signal: Signals) => void;
```

But I write the code like following on purpose to see if I can get any warning,

```
process
  .on('SIGTERM', shutdown)
  .on('warning',shutdown)
  .on('uncaughtException', shutdown);

function shutdown(signal:NodeJS.Signals) {
    console.log(`${signal} bye`)
}
```

tsc compiles without error. Why is that ?

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
in this specific use-case, IMO the `catch-all` function override (the last one) takes precedence:
```ts
on(event: string | symbol, listener: (...args: any[]) => void): this;
```
and `shutdown` matches this function definition. I guess the point here is to allow @types/node work with events with unknown (or future) implementations So it's not closed list and the wrong match fits the catch-all

**Reply by qiulang** [reply]
@peterblazejewicz thanks for replying my question. So now I understand but still, that makes all the other annotations not very useful if I may say so.

**Reply by LinusU** [reply]
The other annotations still gives you autocomplete, which is a nice benefit. If we didn't have the catch-all override, I wouldn't be able to have this code:

```ts
process.on('hello', () => {
  console.log('Hello, World!')
})

process.emit('hello')
```

Now, wether or not this should be allowed could be a valid question. I'm not to familiar with how the `process` object is documented/intended to work, so it's hard to say for me...

**Comment by SimonSchick** [comment]
We always kept specialized event emitters open due to people ~~ab~~using them and then complaining, this is by design.

**Reply by qiulang** [reply]
"The other annotations still gives you autocomplete, which is a nice benefit.“  but actually this is what I got
<img width="913" alt="autocomplete hint" src="https://user-images.githubusercontent.com/516243/178174335-714ce168-cb60-46b3-b966-c06cbd1b64a7.png">

So you can see these are 13 possible hints, no very useful.

**Comment by SimonSchick** [comment]
This particular issue is due is more due to the TS language server not suggesting a good result.
This is an edge case because the constant string used is not identifier as the `NodeJS.Signals` union type.
I'd raise an issue with them, it works for all other instances and `go to definition` also yields the expected navigation even for the signal.
